feat(op-acceptance-tests): cgt

scharissis opened the pull request:
## Description
Adds CGT acceptance tests.

Locally:
<img width="1126" height="202" alt="Screenshot 2025-09-22 at 14 47 03" src="https://github.com/user-attachments/assets/d9d1c12d-f7fb-4f6f-97af-e8cfd1026a4a" />
<img width="1740" height="700" alt="Screenshot 2025-09-22 at 14 47 20" src="https://github.com/user-attachments/assets/2bfeafbb-18df-4cd3-a31f-0aa822020703" />


In CI:
<img width="1737" height="652" alt="Screenshot 2025-09-22 at 16 53 39" src="https://github.com/user-attachments/assets/daf3bdb7-244a-439c-86b7-3c8f62f86fee" />
([source](https://app.circleci.com/pipelines/github/ethereum-optimism/optimism/101531/workflows/831e7bf7-d980-4600-8bda-1610c6dbeb99/jobs/3888883/artifacts))


## Metadata
Intended to be merged into https://github.com/ethereum-optimism/optimism/pull/17412

scharissis committed:
feat(op-acceptance-tests): add acceptance tests for native CGT across L1/L2

- Gate suite via L1Block.isCustomGasToken(); assert name/symbol non-empty
- L2: value transfer pays CGT token (typed amounts via op-service/eth)
- L2: XDM rejects callvalue; L2StandardBridge legacy withdraw reverts
- L1: Portal receive/deposit rejects ETH; introspect SystemConfig addr
- L1: assert SystemConfig.isCustomGasToken() is true

ashitakah commented:
@scharissis Very good work, the tests are solid and cover CGT's main functionality, but if I may suggest, there are three tests I would recommend to ensure that all critical cases are covered. 👍 


```
func TestCGT_PortalReceiveReverts(t *testing.T) {
    // Should verify that direct ETH sends to portal fail in CGT mode
}
```

```
func TestCGT_L1PortalIntrospection(t *testing.T) {
    // Verify that portal returns true when CGT is enabled
}
```

```
func TestCGT_SystemConfigFeatureFlag(t *testing.T) {
    // Verify that feature flag is set correctly
}
```

Then you could make a more complex one with withdrawals and finalizations, but I think that's for another stage.

```
func TestCGT_WithdrawalProveAndFinalize(t *testing.T) {
    // Should test:
    // - proveWithdrawalTransaction() rejects tx with value > 0 in CGT mode
    // - finalizeWithdrawalTransaction() rejects tx with value > 0 in CGT mode  
    // - Withdrawal without value works correctly in CGT mode
}
```

hexshire commented on op-acceptance-tests/tests/custom_gas_token/cgt_introspection_test.go:
```diff
@@ -0,0 +1,62 @@
+	dt.Log("Testing CGT introspection via L1Block")
+
+	token, decimals := ensureCGTOrSkip(t, l2c)
+	if token == ([20]byte{}) {
```
Not sure if this check is needed. Since it is a dummy value.`ensureCGTOrSkip` could just return the `isCustomGasToken` bool, right?

hexshire replied:
Also `gasPayingToken()` is deprecated 

``` solidity
    function gasPayingToken() public pure override returns (address, uint8) {
        revert("L1BlockCGT: deprecated");
    }
 ```

scharissis replied:
Removed that check and deprecated `gasPayingToken()` as per the latest updated to the spec/PR.

scharissis commented:
Thank you @hexshire for the thorough review!
I've pushed an update which hopefully addresses your feedback (except for implementing `TestCGT_WithdrawalProveAndFinalize`), which I may do later or in a follow-up PR.
